Index: usrp/host/lib/inband/symbols_usrp_low_level_cs.h
===================================================================
--- usrp/host/lib/inband/symbols_usrp_low_level_cs.h	(revision 8178)
+++ usrp/host/lib/inband/symbols_usrp_low_level_cs.h	(working copy)
@@ -43,5 +43,6 @@
 static pmt_t s_op_spi_read = pmt_intern("op-spi-read");
 static pmt_t s_op_spi_read_reply = pmt_intern("op-spi-read-reply");
 static pmt_t s_op_delay = pmt_intern("op-delay");
+static pmt_t s_op_mf_set = pmt_intern("s-op-mf-set");
 
 #endif /* INCLUDED_SYMBOLS_USRP_LOW_LEVEL_CS_H */
Index: usrp/host/lib/inband/usrp_inband_usb_packet.cc
===================================================================
--- usrp/host/lib/inband/usrp_inband_usb_packet.cc	(revision 8178)
+++ usrp/host/lib/inband/usrp_inband_usb_packet.cc	(working copy)
@@ -592,6 +592,53 @@
 }
 
 /*!
+ * \brief
+ *
+ * \returns true if the command was added to the packet, false otherwise.
+ */
+bool usrp_inband_usb_packet::cs_mf_set(long threshold, long ncoeff, int32_t *mf_data, size_t data_len)
+{
+  if(!align32())
+    return false;
+
+  int p_len = payload_len();
+
+  int mf_len = data_len + 2 + 4;   // 2 bytes between mbz and addr + 4 for threshold
+
+  int nco_div_6 = ncoeff / 6;
+  int nco_mod_6 = ncoeff % 6;
+  int nco_bits = (nco_div_6<<3) | nco_mod_6;
+  std::cout << "nco_div_6: " << nco_div_6 << "  nco_mod_6: " << nco_mod_6 << "  nco_bits: " << nco_bits << std::endl;
+
+  if((MAX_PAYLOAD - p_len) < (mf_len + CS_FIXED_LEN))
+    return false;
+
+  uint32_t word0 = 0;
+
+  word0 = (
+      ((OP_MF_SET & CS_OPCODE_MASK) << CS_OPCODE_SHIFT)
+    | ((mf_len & CS_LEN_MASK) << CS_LEN_SHIFT)
+    | ((nco_bits & CS_MFNCOEFF_MASK) << CS_MFNCOEFF_SHIFT)
+  );
+
+  uint32_t *payload = (uint32_t *) (d_payload + p_len);
+   *payload = host_to_usrp_u32(word0);
+   
+   payload += 1;
+   *payload = host_to_usrp_u32(threshold);
+
+   // Jump over the first word and write the data
+   // FIXME: Should the data be changed to usrp byte order?
+   payload += 1;
+   memcpy(payload, mf_data, data_len*4);
+
+  // Update payload length
+  incr_header_len(CS_FIXED_LEN + mf_len);
+
+  return true;
+}
+
+/*!
  * \brief Since all control packets contain subpackets which have the length of
  * the subpacket at a uniform location in the subpacket, this will return the
  * subpacket length given a byte offset of the start of the subpacket from the beginning of the packet.
@@ -783,7 +830,12 @@
 
       return pmt_list2(s_op_delay, ticks);
     }
-    
+
+    case OP_MF_SET:
+    {
+      return PMT_NIL;
+    }
+
     default:
       return PMT_NIL;
 
Index: usrp/host/lib/inband/usrp_inband_usb_packet.h
===================================================================
--- usrp/host/lib/inband/usrp_inband_usb_packet.h	(revision 8178)
+++ usrp/host/lib/inband/usrp_inband_usb_packet.h	(working copy)
@@ -50,13 +50,14 @@
     OP_WRITE_REG_MASKED   = 0x03,
     OP_READ_REG           = 0x04,
     OP_READ_REG_REPLY     = 0x05,
-    OP_I2C_WRITE          = 0x06,
+    OP_MF_SET             = 0x06,
     OP_I2C_READ           = 0x07,
     OP_I2C_READ_REPLY     = 0x08,
     OP_SPI_WRITE          = 0x09,
     OP_SPI_READ           = 0x0a,
     OP_SPI_READ_REPLY     = 0x0b,
-    OP_DELAY              = 0x0c
+    OP_DELAY              = 0x0c,
+    OP_I2C_WRITE          = 0x0d
   };
 
   enum flags {
@@ -66,8 +67,9 @@
     FL_START_OF_BURST = 0x10000000,
     FL_END_OF_BURST   = 0x08000000,
     FL_CARRIER_SENSE  = 0x04000000,
+    FL_MATCH_FILTER   = 0x02000000,
 
-    FL_ALL_FLAGS      = 0xfc000000
+    FL_ALL_FLAGS      = 0xff000000
   };
 
   static const int FL_OVERRUN_SHIFT = 31;
@@ -132,6 +134,9 @@
   static const int CS_SPINBYTES_MASK = 0xff;
   static const int CS_SPINBYTES_SHIFT = 24;
 
+  static const int CS_MFNCOEFF_MASK = 0xff;
+  static const int CS_MFNCOEFF_SHIFT = 0;
+
 public:
   
   void set_timestamp(uint32_t timestamp){
@@ -233,6 +238,7 @@
   bool cs_spi_write(long enables, long format, long opt_header_bytes, uint8_t *spi_data, long spi_data_len);
   bool cs_spi_read(long rid, long enables, long format, long opt_header_bytes, long n_bytes);
   bool cs_spi_read_reply(long rid, uint8_t *spi_data, long spi_data_len);
+  bool cs_mf_set(long threshold, long ncoeff, int32_t *mf_data, size_t data_len);
   int cs_len(int payload_offset);
   pmt_t read_subpacket(int payload_offset);
 };
Index: usrp/host/lib/inband/usrp_server.cc
===================================================================
--- usrp/host/lib/inband/usrp_server.cc	(revision 8178)
+++ usrp/host/lib/inband/usrp_server.cc	(working copy)
@@ -30,6 +30,7 @@
 #include <usrp_usb_interface.h>
 #include <fpga_regs_common.h>
 #include <fpga_regs_standard.h>
+#include <complex>
 
 #include <symbols_usrp_server_cs.h>
 #include <symbols_usrp_channel.h>
@@ -40,6 +41,7 @@
 static pmt_t s_shutdown = pmt_intern("%shutdown");
 
 typedef usrp_inband_usb_packet transport_pkt;   // makes conversion to gigabit easy
+typedef std::complex<float>     gr_complex;
 
 const static bool verbose = false;
 
@@ -697,7 +699,7 @@
     return;
 
   // Read information from the properties of the packet
-  bool carrier_sense = false;
+  bool carrier_sense = false, mf_wait = false;
   if(pmt_is_dict(properties)) {
 
     // Check if carrier sense is enabled for the frame
@@ -707,9 +709,16 @@
       if(pmt_eqv(p_carrier_sense, PMT_T)) 
         carrier_sense = true;
     }
+
+    // Check if carrier sense is enabled for the frame
+    if(pmt_t p_mf_wait = pmt_dict_ref(properties, 
+                                      pmt_intern("mf-wait"), 
+                                      PMT_NIL)) {
+      if(pmt_eqv(p_mf_wait, PMT_T)) 
+        mf_wait = true;
+    }
   }
 
-  
   // Determine the number of packets to allocate contiguous memory for
   // bursting over the USB and get a pointer to the memory to be used in
   // building the packets
@@ -727,14 +736,19 @@
       std::min((long)(n_bytes-(n*max_payload_len)), (long)max_payload_len);
   
     if(n == 0) { // first packet gets start of burst flag and timestamp
+
+      int flags = 0;
+
+      flags |= pkts[n].FL_START_OF_BURST;
       
       if(carrier_sense)
-        pkts[n].set_header(pkts[n].FL_START_OF_BURST 
-                           | pkts[n].FL_CARRIER_SENSE, 
-                           channel, 0, payload_len);
-      else
-        pkts[n].set_header(pkts[n].FL_START_OF_BURST, channel, 0, payload_len);
+        flags |= pkts[n].FL_CARRIER_SENSE;
 
+      if(mf_wait) 
+        flags |= pkts[n].FL_MATCH_FILTER;
+
+      pkts[n].set_header(flags, channel, 0, payload_len);
+
       pkts[n].set_timestamp(timestamp);
     
     } else {
@@ -1087,7 +1101,87 @@
       if(verbose)
         std::cout << "[USRP_SERVER] Received SPI read\n";
     }
+    
+    //--------- MF SET -----------//
+    // FIXME: could check that byte count does not exceed 2^8 which
+    // is the max length in the subpacket for # of bytes to read.
+    if(pmt_eq(subp_cmd, s_op_mf_set)) {
+      
+      long threshold = pmt_to_long(pmt_nth(0, subp_data));
+      pmt_t data = pmt_nth(1, subp_data);
 
+      // From pmt_any to gr_complex vector
+      std::vector<gr_complex> coeffs = boost::any_cast<std::vector<gr_complex> > (pmt_any_ref(data));
+
+      // Reverse the coefficients
+      std::reverse(coeffs.begin(), coeffs.end());
+
+      // New data to store the 2-bit coefficient version
+      size_t data_len;
+      pmt_t v_mf_data = pmt_make_u32vector((96*2)/32, 0);
+      int32_t *mf_data = (int32_t *) pmt_u32vector_writeable_elements(v_mf_data, data_len);
+
+      long ncoeff = coeffs.size();
+
+      // Convert from full coefficient set to 2-bit rotated version
+      int32_t *curr_word = mf_data;
+      for(int i=0; i < ncoeff; i++) {
+
+        curr_word = mf_data + i%6;
+
+        float real = coeffs[i].real();
+        float imag = -coeffs[i].imag(); // Need to take the complex conjugate
+
+        int opt_real, opt_imag;
+
+        if(real>0 && imag>=0) {         // Shift Q1 to (0,1) = 1 = 0b01
+          opt_real = 0;
+          opt_imag = 1;
+        }
+        else if(real<=0 && imag>0) {    // Shift Q2 to (-1,0) = 3 = 0b11
+          opt_real = 1;
+          opt_imag = 1;
+        }
+        else if(real<0 && imag<=0) {    // Shift Q3 to (0,-1) = 2 = 0b10
+          opt_real = 1;
+          opt_imag = 0;
+        }
+        else if(real>=0 && imag<0) {    // Shift Q4 to (1,0) = 0 = 0b00
+          opt_real = 0;
+          opt_imag = 0;
+        }
+
+        // This is done for reducing complexity in the FPGA
+        int real_shift = 32-(16-(i/6));
+        int imag_shift = 16-(16-(i/6));
+
+        std::cout << "  optreal: " << opt_real
+                  << "  optimag: " << opt_imag
+                  << std::endl;
+
+        *curr_word |= opt_real << real_shift;
+        *curr_word |= opt_imag << imag_shift;
+        
+      }
+
+
+      // Make the USB packet
+      if(!pkt->cs_mf_set(threshold, ncoeff, mf_data, data_len))
+      {
+        d_cs_usrp->send(s_cmd_usrp_write, 
+                        pmt_list3(invocation_handle, 
+                                  pmt_from_long(channel), 
+                                  v_packet));
+        
+        goto new_packet;
+      }
+      
+      if(verbose)
+        std::cout << "[USRP_SERVER] Received matched filter set\n"
+                  << "  ncoeffs: " << ncoeff
+                  << std::endl;
+    }
+
   subpkt_bail:
     curr_subpkt++;
 
